File: updater/src/main/java/de/darkatra/patcher/updater/service/model/Patch.java
package de.darkatra.patcher.updater.service.model;

import lombok.Data;
import lombok.experimental.Accessors;

import java.util.HashSet;
import java.util.Set;

@Data
@Accessors(chain = true)
public class Patch implements ContextAware {
	private LatestUpdater latestUpdater;
	private Set<ObsoleteFile> obsoleteFiles = new HashSet<>();
	private Set<Packet> packets = new HashSet<>();

	@Override
	public void applyContext(Context context) {
		getLatestUpdater().applyContext(context);
		getPackets().forEach(packet -> packet.applyContext(context));
		getObsoleteFiles().forEach(obsoleteFile -> obsoleteFile.applyContext(context));
	}
}

File: updater/src/main/kotlin/de/darkatra/bfme2/patch/Patch.kt
package de.darkatra.bfme2.patch

data class Patch(
    val packets: Set<Packet>,
    val obsoleteFiles: Set<ObsoleteFile>
) : ContextAware {

    val size: Long = packets.sumOf { packet -> packet.size }
    val compressedSize: Long = packets.sumOf { packet -> packet.compressedSize }

    /**
     * Applies the given [Context] to the [Patch], replacing placeholders of both [Packets][Packet] and [ObsoleteFiles][ObsoleteFile] with their actual value.
     * Ideally this operation should only be performed once per [Patch].
     *
     * @param context the [Context] containing values for placeholders
     */
    override fun applyContext(context: Context) {
        packets.forEach { packet ->
            packet.applyContext(context)
        }
        obsoleteFiles.forEach { obsoleteFile ->
            obsoleteFile.applyContext(context)
        }
    }
}
